docs/DesignDocs/RemoteAttestationCollaterals.md

```diff
@@ -0,0 +1,292 @@
+Remote Attestation Collaterals
+=====
+
+Adding attestation APIs to support collaterals for remote attestation.  Collaterals are information from a second source that the user/challenger can provide for attestation verification.
+
+Motivation
+----------
+
+Currently, the existing collaterals used for Intel SGX quote verification are not exposed to the user.  This makes it difficult for the user/challenger to specify his/her own set of policies.  Adding these new APIs allows the user to specify a validation policy of his/her choosing.  Possible policies:
+1. The user/challenger has the option to specify its set of collaterals during verification.
+2. The user/challenger has the option to provide a datetime to use during verification.  This datetime specified the date and time at which the user wants to do the verification.  If no datetime is provided, it uses the datetime when the collaterals were created.  The user can provide a datetime in the past, enabling auditing of the evidence and collaterals.
+
+
+User Experience
+---------------
+
+There are 3 scenarios.
+
+### 1. Challenger does not specify the collaterals:
+In this scenario the challenger does not receive any collaterals from the attester(where the TEE resides).  The challenger also does not pass any collaterals to the verification function, instead the collaterals are fetched during verification by the OE SDK.
+
+##### Attester generates the evidence (inside an enclave/TEE)
+```C
+// Get OE report
+result = oe_get_report(
+    OE_REPORT_FLAGS_REMOTE_ATTESTATION,
+    NULL, // report_data
+    0,
+    NULL, // opt_params
+    0,
+    &report,
+    report_size);
+```
+
+##### Challenger verifies the evidence (in the untrusted host)
+```C
+...
+
+// Verify report without collaterals
+result = oe_verify_remote_report_with_collaterals(
+            report,
+            report_size,
+            NULL, // collaterals
+            0,    // collaterals_size
+            NULL, // input_validation_time
+            parsed_report);
+...
+```
+
+### 2. Challenger is provided with collaterals:
+In this scenario the attester provides the evidence and collaterals to the challenger.  The challenger then uses these to verify the TEE.
```

Comment:
Do you assume that the collateral provided by the attester is necessarily identical to what the verifier would be able to get itself?  (Even if it is, there's a race condition where it may change right after the attester gets it and before the verifier would ask for it.)   But in general I suspect there might be cases where the verifier might want additional endorsements beyond ones the attester knows about, which scenario 3 admits.  But the point is that the Attester may not know and may send collaterals even though the Verifier is in scenario 3, unless you design a way for a Verifier to tell an Attester to not bother sending collaterals.

So "then uses" might be overstated, maybe you mean "is then free to use" would be more correct.

Reply:
Sounds good.